### PR TITLE
Add a first example of SMPC Tensor

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Add the SMPCTensor Proof of Concept with a dummy example

## How has this been tested?
- There is an example ```test_mpc_tesnsor.py``` - more tests would be added in the following days.
- TODO: Add tests for it.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by iamtrask** [comment]
WHAAA!!!!

File: packages/syft/src/syft/core/node/common/action/run_class_method_action.py [hunk 1]
```diff
@@ -155,7 +155,23 @@ def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
 
             method_name = self.path.split(".")[-1]
 
-            if (
+            if "ShareTensor" in self.path:
```
Comment (iamtrask) [inline packages/syft/src/syft/core/node/common/action/run_class_method_action.py:158]: let's put a TODO comment here in the code to route based on something that partial string search.

File: packages/syft/src/syft/core/node/common/action/smpc_action.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+from ....io.address import Address
+from ....store.storeable_object import StorableObject
+from ....tensor.smpc.share_tensor import ShareTensor
+from ...abstract.node import AbstractNode
+from ...common.action.common import ImmediateActionWithoutReply
+
+MAP_FUNC_TO_NR_GENERATOR_INVOKES = {"__add__": 0, "__mul__": 0}
+
+
+def smpc_add(self_id, other_id, seed, node):
```
Comment (iamtrask) [inline packages/syft/src/syft/core/node/common/action/smpc_action.py:34]: love this

File: packages/syft/src/syft/core/node/common/action/smpc_action.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                ranks_to_run_action=[0],
+                result_id=result_id,
+                address=node.address,
+            )
+        )
+
+    return actions
+
+
+def smpc_mul(self_id, other_id, seed, node):
```
Comment (iamtrask) [inline packages/syft/src/syft/core/node/common/action/smpc_action.py:74]: love this

File: packages/syft/src/syft/core/node/common/action/smpc_action.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        self.args_id = args_id
+        self.kwargs_id = kwargs_id
+        self.id_at_location = result_id
+        self.ranks_to_run_action = ranks_to_run_action
+        self.address = address
+        self.msg_id = msg_id
+        super().__init__(address=address, msg_id=msg_id)
+
+    @staticmethod
+    def filter_actions_after_rank(rank, actions):
```
Comment (iamtrask) [inline packages/syft/src/syft/core/node/common/action/smpc_action.py:136]: great abstraction - i remember talking about this

File: packages/syft/src/syft/core/node/common/action/smpc_action.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        return [
+            action
+            for action in actions
+            if action.ranks_to_run_action == [] or rank in action.ranks_to_run_action
+        ]
+
+    def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
```
Comment (iamtrask) [inline packages/syft/src/syft/core/node/common/action/smpc_action.py:143]: I'm surprised you needed to re-write all this logic. Could you not have simply called the node service which can execute actions?

File: packages/syft/src/syft/core/smpc/test_mpc_tensor.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+vms = [sy.VirtualMachine(name=name) for name in ["alice", "bob", "theo", "andrew"]]
+clients = [vm.get_client() for vm in vms]
+
+# One of the clients hold the secret
+remote_value_1 = clients[0].syft.core.tensor.tensor.Tensor(
+    np.array([[1, 2, 3, 4, -5]], dtype=np.int64)
+)
+
+mpc_tensor_1 = MPCTensor(
```
Comment (iamtrask) [inline packages/syft/src/syft/core/smpc/test_mpc_tensor.py:24]: Let's wrap this constructor in a Tensor.share() method

File: packages/syft/src/syft/core/smpc/test_mpc_tensor.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+print(mpc_tensor_2)
+
+res = mpc_tensor_1 + mpc_tensor_2
+print("Reconstructed value", res.reconstruct())
+
+
+public_value = Tensor(np.array([[3]], dtype=np.int64))
+
+res = mpc_tensor_1 * public_value
+print("Reconstructed value", res.reconstruct())
```
Comment (iamtrask) [inline packages/syft/src/syft/core/smpc/test_mpc_tensor.py:50]: very nice

**Review by iamtrask — APPROVED** [review]
SMPC parts looking really good - added a few comments.

File: packages/syft/src/syft/core/node/common/action/smpc_action.py (hunk 5 above)
Comment (gmuraru) [reply packages/syft/src/syft/core/node/common/action/smpc_action.py:143]: I will check this out :+1:

Comment (gmuraru) [reply packages/syft/src/syft/core/node/common/action/smpc_action.py:143]: I am not sure in this case what the service would be -- I reach this point after:
```RunClassMethodAction``` (here I get all the actions - SMPCActions for the current node) --> I send them to SMPC Service and they execute it (and enter here)

File: packages/syft/src/syft/ast/klass.py [hunk 8]
```diff
@@ -161,6 +161,87 @@ def run_class_method(
...
+
+        Args:
+            *args: Args list of class method.
+            **kwargs: Keyword args of class method.
+
+        Returns:
+            Pointer to object returned by class method.
+        """
+        # stdlib
+        from uuid import UUID
```
Comment (rasswanth-s) [inline packages/syft/src/syft/ast/klass.py:179]: Q: Does it raise circular import error when kept file start?

Comment (gmuraru) [reply packages/syft/src/syft/ast/klass.py:179]: Hmm - it should not

**Pull request closed without merging**